### ISSUE TITLE
Return value of AzureOperationPoller.result( ) 

### ISSUE THREAD

**Issue by SILENCE226** [Azure/azure-sdk-for-python#1003]
I am using SDK: **azure==2.0.0rc6** to do the deployment with DeploymentsOperations.create_or_update() and using AzureOperationPoller.result( ) to get the result every 30 seconds:

```python
result = async_poller.result(timeout=30)
logger.info('Deploy %s provisioning_state: %s', result.name, result.properties.provisioning_state)
```
however, I found the **result** variable sometimes have unexpected content
So I used try catch for it:
```python
try:
  result = async_poller.result(timeout=30)
  logger.info('Deploy %s provisioning_state: %s', result.name, result.properties.provisioning_state)
except AttributeError as e:
  # result variable may not have correct attribute we want
  logger.warning('Getting AttributeError: %s', e)
  logger.debug('result: %s', result)
  logger.debug('response: %s', async_poller._response)
```

here is the log output:
```
2017-02-20 05:27:40 WARNING - Getting AttributeError: 'NoneType' object has no attribute 'provisioning_state'
2017-02-20 05:27:40 DEBUG - result: {'id': None, 'name': None, 'properties': None}
2017-02-20 05:27:40 DEBUG - response: <Response [200]>
```
Some questions about this behavior:
1. From the REST API document: https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-manager-async-operations
   HTTP return code 200 means finished, but before deployment is finished, return code is always 200, is it correct?

2. why the **result** variable contains None type of id/name/properties?

**Comment by huguesv** [comment]
Hi,

Our expert is out on vacation this week, but I have ideas on things you could try.

- I think right now you can get a newer version of azure-mgmt-resource and msrestazure packages than what azure==2.0.0rc6 will give you.  I've heard of bugs being fixed in the async poller, and the history mentions work in that area (https://pypi.python.org/pypi/msrestazure/0.4.7).  I would recommend you only install the latest version of the individual packages you need and check to see if you get the same behavior.

- I'm curious if you get the same behavior if you call wait() instead of result(), and then call get() on DeploymentOperations.

- If you can enable logging of http calls, it might provide a clue as to why the value of properties is None.  I think it can be done this way (from https://github.com/Azure/azure-sdk-for-python/blob/c9d9ada761294bc70c88a31d5166dfa1c149a727/azure-mgmt/tests/mgmt_testcase.py#L22)

```
    import logging
    logger = logging.getLogger('msrest')
    logger.setLevel(logging.DEBUG)
    logger.addHandler(logging.StreamHandler())
```

**Comment by lmazuel** [comment]
Hi @SILENCE226 
I totally agree with the comment of @huguesv (and I'm this expert ;-)).
To answer your most specific question, the async algorithm is more subtle than the HTTP return code. So 200 can be, in some situation, actually correct. As @huguesv said without your log of HTTP calls it's difficult to conclude.
I don't say there is no issue, but this method is used intensively by the [Azure CLI team](https://github.com/Azure/azure-cli) and they don't have problems. So we will have difficulties to reproduce it and your log will be very helpful.

**Comment by SILENCE226** [comment]
@lmazuel 
I'm using **msrestazure** 0.4.4 before and have this issue after ugrade to 0.4.7.
I tried to turn on debug log for "msrest" and found strange thing, please check the logs:

```
2017-02-24T17:10:15.236 2017-02-24 17:10:15,236 [4896:5100] http_logger.py:67 - DEBUG - Response status: 200
2017-02-24T17:10:15.236 2017-02-24 17:10:15,236 [4896:5100] http_logger.py:68 - DEBUG - Response headers:
2017-02-24T17:10:15.251 2017-02-24 17:10:15,236 [4896:5100] http_logger.py:70 - DEBUG -     'Cache-Control': 'no-cache'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:70 - DEBUG -     'Pragma': 'no-cache'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:70 - DEBUG -     'Content-Type': 'application/json; charset=utf-8'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:70 - DEBUG -     'Content-Encoding': 'gzip'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:70 - DEBUG -     'Expires': '-1'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:70 - DEBUG -     'Vary': 'Accept-Encoding'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:70 - DEBUG -     'x-ms-ratelimit-remaining-subscription-reads': '14958'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:70 - DEBUG -     'x-ms-request-id': '6799b30b-7992-40dc-83de-6f63a3c92c88'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:70 - DEBUG -     'x-ms-correlation-request-id': '6799b30b-7992-40dc-83de-6f63a3c92c88'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:70 - DEBUG -     'x-ms-routing-request-id': 'CENTRALUS:20170224T171015Z:6799b30b-7992-40dc-83de-6f63a3c92c88'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:70 - DEBUG -     'Strict-Transport-Security': 'max-age=31536000; includeSubDomains'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:70 - DEBUG -     'Date': 'Fri, 24 Feb 2017 17:10:14 GMT'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:70 - DEBUG -     'Content-Length': '140'
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:73 - DEBUG - Response content:
2017-02-24T17:10:15.251 2017-02-24 17:10:15,251 [4896:5100] http_logger.py:87 - DEBUG - b'{"status":"Running"}'
2017-02-24T17:10:38.776 2017-02-24 17:10:38,776 [4896:2192] preallocate_commands.py:861 - WARNING - Getting AttributeError: 'NoneType' object has no attribute 'provisioning_state'
2017-02-24T17:10:38.776 2017-02-24 17:10:38,776 [4896:2192] preallocate_commands.py:862 - DEBUG - result: {'properties': None, 'name': None, 'id': None}
2017-02-24T17:10:38.776 2017-02-24 17:10:38,776 [4896:2192] preallocate_commands.py:863 - DEBUG - response: <Response [200]>
```
The response body has JSON
```JSON
'{"status":"Running"}'
```
but from the async_poller.result() still get nothing of provisioning_state ( please checking the last 3 lines of log)
after digging into the code of msrestazure, I found [this line](https://github.com/Azure/msrestazure-for-python/blob/master/msrestazure/azure_operation.py#L312) will replace **status**  variable after v0.4.5

my question is:
is it correct to keep checking deployment status by result()?

```python
while True:
  try:
    result = async_poller.result(timeout=30)
    logger.info('Deploy %s provisioning_state: %s', result.name, result.properties.provisioning_state)
    if result.properties.provisioning_state == 'Succeeded':
       break
  except AttributeError as e:
    # result variable may not have correct attribute we want
    logger.warning('Getting AttributeError: %s', e)
    logger.debug('result: %s', result)
    logger.debug('response: %s', async_poller._response)
```

if not correct, any suggestions to get the correct deployment status?

Thanks!

**Comment by lmazuel** [comment]
Hi @SILENCE226 

This is a very interesting scenario. Our purpose is to hide the mechanics to help the user. By looking directly into `provisionning_state`, you use a side-effect unexpected.

Actually, if you look at the async algorithm you link, Deployment is using the `Azure-AsyncOperation` header, so polling will be `{"status":"Running"}` (as you saw in the log). So the actual *current* result is this status. We clearly fail to parse it correctly returning an empty id/name/properties instance (base Azure Model by the way), but it's another issue I have to fix (you should get `{"status":"Running"}` as a dict). We never meant to actually return the current resource at this state, so the code you wrote in `msrestazure` < 0.4.7 was working, but not on purpose. Actually, you were getting the initial response forever, until the call was done and this response was replaced (so it was not the current state anyway).

If you look at the full algorithm, you will see that there is three async algorithms: `Azure-AsyncOperation`, `Location` and direct resource (in that order of priority). To avoid having the user the manually handle this, the poller hides that so you can write this (example):

```python
while True:
    result = poller.result(timeout=40)
    if poller.done():
        break
    print("Not done yet")
print(result.name)
```

This is working whatever the async approach the service team takes (even if they are not using provisionningState, since it's possible).

Considering that you know for sure Deployment is using provisionningState, if you still want to print the str, you can do
```python
while True:
    result = poller.result(timeout=40)
    if poller.done():
        break
    deployment = client.deployments.get(rg_name, deployment_name)
    print("Current state: "+deployment.properties.provisionning_state)
print(result.name)
```

Does this make sense?

**Comment by SILENCE226** [comment]
Hi @lmazuel 

Thanks for detailed explanation!
I understand your design purpose now.

May I have a request to add new API to get the status string?
as I know, this string is stored in ```AzureOperationPoller._operation.status```

**Comment by lmazuel** [comment]
Hi @SILENCE226 , I created two issues in the repo with the poller code:
https://github.com/Azure/msrestazure-for-python/issues/19 : the empty model as intermediate result call
https://github.com/Azure/msrestazure-for-python/issues/18 : your status getter

I will try to add that in the next iteration, should be possible since it's easy to do.

Thank you for your report! If you think we solved the initial question, maybe we can close the issue?
